driver/docker/docker.go

```diff
@@ -54,34 +52,14 @@ func (d *Driver) AddConfigurationOptions(opts ...ConfigurationOption) {
 
 // GetContainerConfig returns a copy of the container configuration
 // used by the driver during container exec
-func (d *Driver) GetContainerConfig() (container.Config, error) {
-	cpy, err := copystructure.Copy(*d.containerCfg)
-	if err != nil {
-		return container.Config{}, err
-	}
-
-	containerCfg, ok := cpy.(container.Config)
-	if !ok {
-		return container.Config{}, errors.New("unable to process container config")
-	}
-
-	return containerCfg, nil
+func (d *Driver) GetContainerConfig() container.Config {
+	return d.containerCfg
 }
 
 // GetContainerHostConfig returns a copy of the container host configuration
 // used by the driver during container exec
-func (d *Driver) GetContainerHostConfig() (container.HostConfig, error) {
-	cpy, err := copystructure.Copy(*d.containerHostCfg)
-	if err != nil {
-		return container.HostConfig{}, err
-	}
-
-	hostCfg, ok := cpy.(container.HostConfig)
-	if !ok {
-		return container.HostConfig{}, errors.New("unable to process container host config")
-	}
-
-	return hostCfg, nil
+func (d *Driver) GetContainerHostConfig() container.HostConfig {
+	return d.containerHostCfg
```

Comment:
There is a difference between what you had before and what this does here.

Previously, your code made a deep copy of the entire config, ensuring that modifying the returned config could not affect the config used in `exec`. In this new function, it makes a copy but not how you may expect, and the calling code could use it to change the config, for example changing the data in the backing array behind the `CapAdd` slice.

```golang
// Assume that CapAdd already had SYS_PTRACE, the caller could alter it to be SYS_ADMIN instead without going through our defined interface of AddConfigurationOptions
hostCfg1 := d.GetContainerHostConfig()
hostCfg1.CapAdd[0] = "SYS_ADMIN"
hostCfg2 := d.GetContainerHostConfig()
is.Equal(hostCfg1, hostCfg2)
```

This is a contrived example, but is something that sometimes comes up when returning a slice or value that contains a slice, and then later we are wondering why the contents of the _original_ slice have changed inadvertently. It can be hard to track down, but doing a deep copy can avoid stuff like this. It's something that we should think about if we want to remove that protection or if we don't think it matters for this use case / library. 

Maybe we care, maybe we don't. 😀 I wanted to make sure you understood the difference in behavior between the two pieces of code. Otherwise previously if you had only cared about converting from a pointer to a value, there were cheaper methods so I had assumed you cared about solving this problem too.

Reply:
Thank you.  I had the misunderstanding that by changing the field to a value, the deep copy was no longer applicable.  You've helped me see that a deep copy is applicable whenever the field struct itself contains slice fields, regardless of if we access by value or by pointer.  Therefore, I've updated to maintain the deep copy logic when accessing, as I think this is the prudent way to go, for reasons you've illustrated!